File: Bomberman/application/modules/Mediator.js
class Mediator {

    constructor(options) {
        // triggers
        this.TRIGGERS = options.TRIGGERS;
        this.triggers = {};
        for (let key in this.TRIGGERS) {
            this.triggers[key] = function () { return null; };
        }
        // events
        this.EVENTS = options.EVENTS;
        this.events = {};
        for (let key in this.EVENTS) {
            this.events[key] = [];
        }
    }

    // вернуть типы триггеров
    getTriggers() {
        return this.TRIGGERS;
    }
    // установить триггер
    set(name, func) {
        if (name && func instanceof Function) {
            this.triggers[name] = func;
        }
    }
    // вызвать триггер
    get(name, data) {
        if (this.triggers[name] instanceof Function) {
            return this.triggers[name](data);
        }
        return null;
    }

    // вернуть типы событий
    getEvents() {
        return this.EVENTS;
    }
    // подписаться на событие
    subscribe(name, func) {
        if (name && func instanceof Function) {
            this.events[name].push(func);
        }
    }
    // вызвать событие
    call(name, data) {
        if (this.events[name] && this.events[name].length) {
            for (let i = 0; i < this.events[name].length; i++) {
                if (this.events[name][i] instanceof Function) {
                    this.events[name][i](data);
                }
            }
        }
    }
}
module.exports = Mediator;

File: SmartHouse/application/modules/Mediator.py
class Mediator:
    EVENT_TYPES = {} # типы событий
    events = {} # списки событий
    TRIGGER_TYPES = {} # типы триггеров
    triggers = {} # список триггеров по их типам

    def __init__(self, options):
        self.EVENT_TYPES = options['EVENTS']
        self.TRIGGER_TYPES = options['TRIGGERS']
        for key in self.EVENT_TYPES.keys():
            self.events.update({ self.EVENT_TYPES[key]: [] })
        for key in self.TRIGGER_TYPES.keys():
            self.triggers.update({ self.TRIGGER_TYPES[key]: lambda x = None: x })

    def __del__(self):
        self.events.clear()
        self.triggers.clear()

    def getEventTypes(self):
        return self.EVENT_TYPES

    def getTriggerTypes(self):
        return self.TRIGGER_TYPES

    def subscribe(self, name, func):
        if name and callable(func):
            self.events.get(name).append(func)

    def call(self, name, data = None):
        if (name):
            cbs = self.events.get(name)
            if cbs:
                for cb in cbs:
                    if callable(cb):
                        if data:
                            cb(data)
                        else:
                            cb()
    
    def set(self, name, func):
        if name and callable(func):
            self.triggers.update({ name: func })

    def get(self, name, data = None):
        if (name):
            cb = self.triggers.get(name)
            if cb and callable(cb):
                if data:
                    return cb(data)
                return cb()
        return None
